### ISSUE TITLE
Training loss history

### ISSUE THREAD

**Issue by alibabadoufu** [facebookresearch/swav#12]
Hi,

Thank you so much for sharing your codes!

May I know if you have a copy of your loss record?

When I trained your model from scratch, the loss was stacked around 8 for the first 2 epochs. (I am still training the model)

Is it the same for you?

Thank you.

**Comment by alibabadoufu** [comment]
For example, from epoch 1 onwards, the loss has been stuck around 8. Is this normal?

```
INFO - 08/11/20 13:38:30 - 1 day, 22:12:26 - ============ Starting epoch 163 ... ============
INFO - 08/11/20 13:38:32 - 1 day, 22:12:28 - Epoch: [163][0]    Time 1.958 (1.958)      Data 1.576 (1.576)      Loss 8.0066 (8.0066)    Lr: 3.1982
INFO - 08/11/20 13:39:16 - 1 day, 22:13:13 - Epoch: [163][50]   Time 0.890 (0.919)      Data 0.000 (0.031)      Loss 8.0065 (8.0066)    Lr: 3.1976
INFO - 08/11/20 13:40:01 - 1 day, 22:13:58 - Epoch: [163][100]  Time 0.894 (0.909)      Data 0.000 (0.016)      Loss 8.0068 (8.0066)    Lr: 3.1970
INFO - 08/11/20 13:40:46 - 1 day, 22:14:43 - Epoch: [163][150]  Time 0.897 (0.905)      Data 0.000 (0.011)      Loss 8.0066 (8.0066)    Lr: 3.1964
INFO - 08/11/20 13:41:31 - 1 day, 22:15:28 - Epoch: [163][200]  Time 0.892 (0.904)      Data 0.000 (0.008)      Loss 8.0065 (8.0066)    Lr: 3.1958
INFO - 08/11/20 13:42:16 - 1 day, 22:16:13 - Epoch: [163][250]  Time 0.910 (0.902)      Data 0.000 (0.006)      Loss 8.0068 (8.0066)    Lr: 3.1953
INFO - 08/11/20 13:43:01 - 1 day, 22:16:58 - Epoch: [163][300]  Time 0.899 (0.902)      Data 0.000 (0.005)      Loss 8.0066 (8.0066)    Lr: 3.1947
INFO - 08/11/20 13:43:46 - 1 day, 22:17:42 - Epoch: [163][350]  Time 0.892 (0.901)      Data 0.000 (0.005)      Loss 8.0065 (8.0066)    Lr: 3.1941
INFO - 08/11/20 13:44:31 - 1 day, 22:18:27 - Epoch: [163][400]  Time 0.897 (0.901)      Data 0.000 (0.004)      Loss 8.0068 (8.0066)    Lr: 3.1935
INFO - 08/11/20 13:45:16 - 1 day, 22:19:12 - Epoch: [163][450]  Time 0.901 (0.900)      Data 0.000 (0.004)      Loss 8.0066 (8.0066)    Lr: 3.1929
INFO - 08/11/20 13:46:00 - 1 day, 22:19:57 - Epoch: [163][500]  Time 0.891 (0.900)      Data 0.000 (0.003)      Loss 8.0065 (8.0066)    Lr: 3.1923
INFO - 08/11/20 13:46:45 - 1 day, 22:20:42 - Epoch: [163][550]  Time 0.922 (0.900)      Data 0.000 (0.003)      Loss 8.0068 (8.0066)    Lr: 3.1918
INFO - 08/11/20 13:47:30 - 1 day, 22:21:27 - Epoch: [163][600]  Time 0.889 (0.900)      Data 0.000 (0.003)      Loss 8.0066 (8.0066)    Lr: 3.1912
INFO - 08/11/20 13:48:15 - 1 day, 22:22:12 - Epoch: [163][650]  Time 0.897 (0.899)      Data 0.000 (0.003)      Loss 8.0065 (8.0066)    Lr: 3.1906
INFO - 08/11/20 13:49:00 - 1 day, 22:22:57 - Epoch: [163][700]  Time 0.897 (0.900)      Data 0.000 (0.002)      Loss 8.0068 (8.0066)    Lr: 3.1900
INFO - 08/11/20 13:49:44 - 1 day, 22:23:41 - Epoch: [163][750]  Time 0.897 (0.898)      Data 0.000 (0.002)      Loss 8.0066 (8.0066)    Lr: 3.1894
INFO - 08/11/20 13:50:29 - 1 day, 22:24:26 - Epoch: [163][800]  Time 0.896 (0.898)      Data 0.000 (0.002)      Loss 8.0065 (8.0066)    Lr: 3.1888
INFO - 08/11/20 13:51:14 - 1 day, 22:25:11 - Epoch: [163][850]  Time 0.893 (0.898)      Data 0.000 (0.002)      Loss 8.0068 (8.0066)    Lr: 3.1883
INFO - 08/11/20 13:51:59 - 1 day, 22:25:55 - Epoch: [163][900]  Time 0.900 (0.898)      Data 0.000 (0.002)      Loss 8.0066 (8.0066)    Lr: 3.1877
INFO - 08/11/20 13:52:44 - 1 day, 22:26:40 - Epoch: [163][950]  Time 0.911 (0.898)      Data 0.000 (0.002)      Loss 8.0065 (8.0066)    Lr: 3.1871
INFO - 08/11/20 13:53:00 - 1 day, 22:26:57 - Epoch: [163][1000] Time 0.307 (0.870)      Data 0.000 (0.002)      Loss 8.0068 (8.0066)    Lr: 3.1865
INFO - 08/11/20 13:53:17 - 1 day, 22:27:14 - Epoch: [163][1050] Time 0.303 (0.845)      Data 0.000 (0.002)      Loss 8.0066 (8.0066)    Lr: 3.1859
INFO - 08/11/20 13:53:35 - 1 day, 22:27:31 - Epoch: [163][1100] Time 0.303 (0.822)      Data 0.000 (0.002)      Loss 8.0065 (8.0066)    Lr: 3.1853
INFO - 08/11/20 13:54:02 - 1 day, 22:27:59 - Epoch: [163][1150] Time 0.904 (0.810)      Data 0.000 (0.002)      Loss 8.0068 (8.0066)    Lr: 3.1848
INFO - 08/11/20 13:54:47 - 1 day, 22:28:43 - Epoch: [163][1200] Time 0.887 (0.814)      Data 0.000 (0.001)      Loss 8.0066 (8.0066)    Lr: 3.1842
INFO - 08/11/20 13:55:28 - 1 day, 22:29:25 - Epoch: [163][1250] Time 0.811 (0.814)      Data 0.000 (0.001)      Loss 8.0065 (8.0066)    Lr: 3.1836
INFO - 08/11/20 13:56:11 - 1 day, 22:30:07 - Epoch: [163][1300] Time 0.886 (0.816)      Data 0.000 (0.001)      Loss 8.0068 (8.0066)    Lr: 3.1830
INFO - 08/11/20 13:56:56 - 1 day, 22:30:52 - Epoch: [163][1350] Time 0.898 (0.819)      Data 0.000 (0.001)      Loss 8.0066 (8.0066)    Lr: 3.1824
INFO - 08/11/20 13:57:40 - 1 day, 22:31:37 - Epoch: [163][1400] Time 0.915 (0.821)      Data 0.000 (0.001)      Loss 8.0065 (8.0066)    Lr: 3.1818
```

**Comment by xiaoyu825** [comment]
I also encountered a similar problem，i used this [version()](https://github.com/facebookresearch/swav/blob/master/scripts/swav_100ep_pretrain.sh).My loss stucked around 7+。and i 
 use the model making a Linear classification on ImageNet only get top1 acc around 31%.

**Comment by mathildecaron31** [comment]
Hi @alibabadoufu, @xiaoyu825, thanks for your interest in our work.

I've also observed in some of my trainings that the loss gets stuck at ln(K) (8.0066 for K=3000) if the hyperparameters are not set properly. In that case, the model does not train and it is expected to have a poor value when evaluating it.
Can you copy-paste the arguments you are using for this run?
By the way, in my experience if you see that the loss is stuck for the first epochs then you can completely stop the training, no need to let it run until 160 epochs.

@xiaoyu825, are you using *exactly* all the arguments from this [script](https://github.com/facebookresearch/swav/blob/master/scripts/swav_100ep_pretrain.sh) ?

**Comment by mathildecaron31** [comment]
On Fig 7 of the paper, you can see an example of the loss we typically observe in our swav trainings. It starts around 8 and decreases until ~3

**Comment by alibabadoufu** [comment]
> Hi @alibabadoufu, @xiaoyu825, thanks for your interest in our work.
> 
> I've also observed in some of my trainings that the loss gets stuck at ln(K) (8.0066 for K=3000) if the hyperparameters are not set properly. In that case, the model does not train and it is expected to have a poor value when evaluating it.
> Can you copy-paste the arguments you are using for this run?
> By the way, in my experience if you see that the loss is stuck for the first epochs then you can completely stop the training, no need to let it run until 160 epochs.
> 
> @xiaoyu825, are you using _exactly_ all the arguments from this [script](https://github.com/facebookresearch/swav/blob/master/scripts/swav_100ep_pretrain.sh) ?

Thanks for your reply. 

Do you know exactly which hyperparameter is causing this problem?

**Comment by xiaoyu825** [comment]
Thanks for your reply.
we don't have enough GPUs，so we add accumulating gradients to make sure the batchsize 4096 in two v100s with all 8gpus,so changes in [main_swav.py](https://github.com/facebookresearch/swav/blob/139623b23be7d1f0caeea7aa5fb292f811df22d7/main_swav.py#L316)
 `        accumulation_steps = 8
        loss = loss/accumulation_steps  
        # ============ backward and optim step ... ============
        #optimizer.zero_grad()
        if args.use_fp16:
            with apex.amp.scale_loss(loss, optimizer) as scaled_loss:
                scaled_loss.backward()
        else:
            loss.backward()
        # cancel some gradients
        if iteration < args.freeze_prototypes_niters:
            for name, p in model.named_parameters():
                if "prototypes" in name:
                    p.grad = None
        if((it+1)%accumulation_steps)==0:
            optimizer.step()
            optimizer.zero_grad()`

and, here is my script:
`python -m torch.distributed.launch --nproc_per_node=4 --nnodes=2 --node_rank=0 --master_addr="10.207.43.19" --master_port=23459 main_swav.py \
--data_path /home/data/imagenet \
--dist_url env:// \
--nmb_crops 2 6 \
--size_crops 224 96 \
--min_scale_crops 0.14 0.05 \
--max_scale_crops 1. 0.14 \
--crops_for_assign 0 1 \
--temperature 0.1 \
--epsilon 0.05 \
--sinkhorn_iterations 3 \
--feat_dim 128 \
--nmb_prototypes 3000 \
--queue_length 0 \
--epochs 100 \
--batch_size 64 \
--base_lr 4.8 \
--final_lr 0.0048 \
--freeze_prototypes_niters 313 \
--wd 0.000001 \
--warmup_epochs 10 \
--start_warmup 0.3 \
--arch resnet50 \
--use_fp16 true \
--sync_bn apex \
--checkpoint_freq 1 \
--dump_path "./experiments/swav_100ep_pretrain"`

**Comment by xiaoyu825** [comment]
Thanks for your reply.
we don't have enough GPUs，so we add accumulating gradients to make sure the batchsize 4096 in two v100s with all 8gpus,so changes in [main_swav.py](https://github.com/facebookresearch/swav/blob/139623b23be7d1f0caeea7aa5fb292f811df22d7/main_swav.py#L316)

>      
        accumulation_steps = 8
        loss = loss/accumulation_steps  
        # ============ backward and optim step ... ============
        #optimizer.zero_grad()
        if args.use_fp16:
            with apex.amp.scale_loss(loss, optimizer) as scaled_loss:
                scaled_loss.backward()
        else:
            loss.backward()
        # cancel some gradients
        if iteration < args.freeze_prototypes_niters:
            for name, p in model.named_parameters():
                if "prototypes" in name:
                    p.grad = None
        if((it+1)%accumulation_steps)==0:
            optimizer.step()
            optimizer.zero_grad()

and, here is my script:

> python -m torch.distributed.launch --nproc_per_node=4 --nnodes=2 --node_rank=0 --master_addr="10.207.43.19" --master_port=23459 main_swav.py \
--data_path /home/data/imagenet \
--dist_url env:// \
--nmb_crops 2 6 \
--size_crops 224 96 \
--min_scale_crops 0.14 0.05 \
--max_scale_crops 1. 0.14 \
--crops_for_assign 0 1 \
--temperature 0.1 \
--epsilon 0.05 \
--sinkhorn_iterations 3 \
--feat_dim 128 \
--nmb_prototypes 3000 \
--queue_length 0 \
--epochs 100 \
--batch_size 64 \
--base_lr 4.8 \
--final_lr 0.0048 \
--freeze_prototypes_niters 313 \
--wd 0.000001 \
--warmup_epochs 10 \
--start_warmup 0.3 \
--arch resnet50 \
--use_fp16 true \
--sync_bn apex \
--checkpoint_freq 1 \
--dump_path "./experiments/swav_100ep_pretrain"

**Comment by alibabadoufu** [comment]
This is my script:

```
--data_path ~/datasets/ILSVRC12 \
--epochs 100 \
--base_lr 4.8 \
--final_lr 0.0048 \
--warmup_epochs 10 \
--start_warmup 0.3 \
--batch_size 64 \
--wd 0.000001 \
--size_crops 224 96 \
--nmb_crops 2 6 \
--min_scale_crops 0.14 0.05 \
--max_scale_crops 1. 0.14 \
--use_fp16 true \
--temperature 0.1 \
--epsilon 0.05 \
--sinkhorn_iterations 3 \
--crops_for_assign 0 1 \
--nmb_prototypes 3000 \
--freeze_prototypes_niters 313 \
--queue_length 0 \
--feat_dim 128 \
--epoch_queue_starts 15 \
```

**Comment by mathildecaron31** [comment]
> Do you know exactly which hyperparameter was causing this problem?
Several hyper-parameters can be the root of the problem. For example, my loss was stuck at ln(K) when using epsilon too high (>0.1). When using multi-crop I had to freeze the prototypes at the beginning (freeze_prototypes_niters 313) otherwise the loss gets stuck. Also I've observed that removing the batch-normalization from the head sometimes helps the optimization.

@alibabadoufu are you launching this script on a batch size of 4096 ? If yes, I'm very surprised that it does not optimize since it is exactly my set of hyperparameters :/

@xiaoyu825 I have not experimented at all with accumulated gradients so I'm afraid I cannot really help you. Maybe you would better train at smaller batch sizes.

**Comment by etema19** [comment]
@alibabadoufu Have you fixed it?

**Comment by wuziheng** [comment]
> I changed the setting to 200 epoch version, everything goes well !

may i ask the final loss and top1 of your 200epoch experiments.?

**Comment by wuziheng** [comment]
> On Fig 7 of the paper, you can see an example of the loss we typically observe in our swav trainings. It starts around 8 and decreases until ~3

hello, myfriend, I reimplement this swav in another SelfSup code base, almost copy/ paste. Now I got stuck at a strange situation， which my loss starts around 8 , decreases until ~4.4 at 100epoch , but linear_eval got top1~=0. this linear_eval config is allright with swav official model which got 73.5%. Any advise or deep communication for debug this, thankyou !

**Comment by mathildecaron31** [comment]
https://github.com/facebookresearch/swav#the-loss-does-not-decrease-and-is-stuck-at-lnnmb_prototypes-8006-for-3000-prototypes
